File: 01-beginner/03-services-and-actions/demo/python/add_two_ints_server.py
#!/usr/bin/env python3
"""A service server that adds two integers on request."""

import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts


class AddTwoIntsServer(Node):
    def __init__(self):
        super().__init__("add_two_ints_server")

        # create_service(srv_type, service_name, callback). Unlike a
        # subscription callback, this callback both receives an argument
        # (the request) AND must return a value (the response) — the
        # client is waiting on that return value.
        self._service = self.create_service(
            AddTwoInts, "add_two_ints", self._handle_request
        )
        self.get_logger().info("add_two_ints service ready")

    def _handle_request(
        self, request: AddTwoInts.Request, response: AddTwoInts.Response
    ) -> AddTwoInts.Response:
        # ROS2 service callbacks are handed an already-constructed
        # response object to fill in and return, rather than constructing
        # a new one — this is the rclpy convention for service callbacks.
        response.sum = request.a + request.b
        self.get_logger().info(
            f"Incoming request: {request.a} + {request.b} = {response.sum}"
        )
        return response


def main():
    rclpy.init()
    node = AddTwoIntsServer()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: 01-beginner/03-services-and-actions/demo/cpp/src/add_two_ints_server.cpp
// A service server that adds two integers on request.
// C++ equivalent of demo/python/add_two_ints_server.py.

#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

using AddTwoInts = example_interfaces::srv::AddTwoInts;

class AddTwoIntsServer : public rclcpp::Node
{
public:
  AddTwoIntsServer() : Node("add_two_ints_server")
  {
    // create_service takes the service name and a callback that receives
    // shared_ptrs to the request and response — in rclcpp, unlike rclpy,
    // the response is filled in through a pointer rather than returned.
    service_ = this->create_service<AddTwoInts>(
      "add_two_ints",
      std::bind(&AddTwoIntsServer::handle_request, this,
        std::placeholders::_1, std::placeholders::_2));
    RCLCPP_INFO(this->get_logger(), "add_two_ints service ready");
  }

private:
  void handle_request(
    const std::shared_ptr<AddTwoInts::Request> request,
    std::shared_ptr<AddTwoInts::Response> response)
  {
    response->sum = request->a + request->b;
    RCLCPP_INFO(this->get_logger(), "Incoming request: %ld + %ld = %ld",
      request->a, request->b, response->sum);
  }

  rclcpp::Service<AddTwoInts>::SharedPtr service_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<AddTwoIntsServer>());
  rclcpp::shutdown();
  return 0;
}
